Fix #18481: Making correctness feedback the default setting for all lessons

AFZL210 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #18481 .
2. This PR does the following: It updates the code related to exploration creation and updating.

Updated the storage layer exploration model correctness_feedback field default value to true and removed correctness_feedback argument from save_new_valid_exploration(), create_default_exploration(), save_new_linear_exp_with_state_names_and_interactions(), update_exploration(), and update_correctness_feedback_enabled() functions and their references.

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...",
followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] The linter/Karma presubmit checks have passed on my local machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
- [x] My PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] I have assigned the correct reviewers to this PR (or will leave a
comment with the phrase "@{{reviewer_username}} PTAL" if I don't have
permissions to assign reviewers directly).


## Proof that changes are correct


![286863763-51929a90-e68d-467d-81b6-0d6c2657694c](https://github.com/oppia/oppia/assets/79896602/f74be5df-1643-421e-b3f1-5c0432e92131)

[Screencast from 16-12-23 07:45:46 PM IST.webm](https://github.com/oppia/oppia/assets/79896602/5956d8cc-a994-41ca-a218-99cb3a22a11a)



![Screenshot from 2023-12-16 19-46-02](https://github.com/oppia/oppia/assets/79896602/79cfdffc-fc9a-4fb4-8a35-98b38c774a93)




<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface. Please also include
videos/screenshots of the developer tools browser console, so that we can be
sure that there are no console errors.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

N/A

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone
N/A

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language
N/A

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

AFZL210 committed:
Fix#18481: Making correctness feedback the default setting for all lessons

AFZL210 committed:
remove correctness_feedback from exp_domain

AFZL210 committed:
remove correctness_feedback from backend

AFZL210 committed:
remove unused code

AFZL210 commented:
@seanlip PTAL

seanlip commented on core/storage/exploration/gae_models.py:
```diff
@@ -232,8 +232,9 @@ class ExplorationModel(base_models.VersionedModel):
         datastore_services.BooleanProperty(default=True, indexed=True))
     # A boolean indicating whether correctness feedback is enabled in this
     # exploration.
+    # DEPRECATED: Enable correctness feedback for all lessons.
```
Move this to the bottom of the list of properties. See how we do it in 
core/storage/email/gae_models.py -- you should have a similar format and docstring here. Its value doesn't matter any more.

You don't need to make it default=True either since we shouldn't be reading it. When reading the storage model to the domain layer we should ignore the deprecated field entirely.

AFZL210 replied:
Done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -3306,7 +3303,7 @@ def save_new_valid_exploration(
 
         exploration.objective = objective
-        exploration.correctness_feedback_enabled = correctness_feedback_enabled
+        exploration.correctness_feedback_enabled = feconf.DEFAULT_CORRECTNESS_FEEDBACK_ENABLED # pylint: disable=line-too-long
```
This should not be necessary at all. Ditto below.

AFZL210 replied:
removed it.

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -6575,7 +6545,7 @@ def __init__(
         self.param_changes = param_changes
         self.auto_tts_enabled = auto_tts_enabled
-        self.correctness_feedback_enabled = correctness_feedback_enabled
+        self.correctness_feedback_enabled = feconf.DEFAULT_CORRECTNESS_FEEDBACK_ENABLED # pylint: disable=line-too-long
```
You should be dropping the entire field and property from the domain layer (and pretty much all layers). Ditto throughout this PR.

We just need to make sure that items that include correctness_feedback_enabled still parse correctly. These items might be in the datastore (it should be fine if you never read the field when fetching from the datastore and converting it into a domain object), the memcache (we should again ignore that field when we convert to a domain object), and localStorage (I don't think we store this info there though).

AFZL210 replied:
Done

seanlip reviewed (changes requested):
Thanks @AFZL210. I think this doesn't fully remove correctness_feedback_enabled -- left some inline comments.

AFZL210 committed:
remove correctness_feedback_enabled from backend

AFZL210 committed:
update frontend

AFZL210 committed:
remove correctness feedback from fronend interfaces

AFZL210 committed:
fix lint issues

AFZL210 committed:
fix lint issues

AFZL210 committed:
removed unused code

AFZL210 committed:
fix linter

AFZL210 committed:
undo batch job changes

AFZL210 committed:
fix backend test

AFZL210 committed:
fix job

AFZL210 commented:
@seanlip PTAL, removed it from all layers from the backend and also updated the frontend interfaces to accommodate this change. Now, we do not read the correctness_feedback_enabled property from the datastore, and the banner is always shown when the answer is correct.

DubeySandeep commented on core/tests/webdriverio_utils/workflow.js:
```diff
@@ -209,16 +208,12 @@ var createAndPublishTwoCardExploration = async function(
     await explorationEditorSettingsTab.setLanguage(language);
   }
-  if (!correctnessFeedbackIsEnabled) {
-    await explorationEditorSettingsTab.disableCorrectnessFeedback();
```
Can we remove`disableCorrectnessFeedback` from explorationEditorSettingsTab?

AFZL210 replied:
@DubeySandeep indeed, we can remove this as it's not being used anywhere now.

DubeySandeep reviewed (changes requested):
@AFZL210 Thanks for the PR, left one comment for the few codeowner files PTAL!

AFZL210 committed:
removed disableCorrectnessFeedback button test

AFZL210 commented:
@seanlip @DubeySandeep PTAL

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -838,17 +838,6 @@ class EditExplorationPropertyAutoTtsEnabledCmd(ExplorationChange):
     old_value: bool
 
 
-class EditExplorationPropertyCorrectnessFeedbackEnabledCmd(ExplorationChange):
```
You might need to keep this so that we can interpret historical data. Do we have an existing standard/approach for deprecating old commit-cmd types?

AFZL210 replied:
We can keep this, but it won't be used as we have removed the option to edit correctness feedback. I tried finding PR/issues related to CMD but didn't find anything related. Let me know if there's some better way to handle this.

seanlip replied:
See e.g. STATE_PROPERTY_WRITTEN_TRANSLATIONS_DEPRECATED or STATE_PROPERTY_NEXT_CONTENT_ID_INDEX_DEPRECATED (and look at the PRs affecting that, which you can find using the "blame" functionality in github) for how this has been handled in the past. You'll want to follow a similar approach here.

AFZL210 replied:
Done.  have followed the approach used to remove the next_content_id state property.

seanlip commented on core/storage/exploration/gae_models.py:
```diff
@@ -230,17 +230,18 @@ class ExplorationModel(base_models.VersionedModel):
+    # exploration.
+    # DEPRECATED. Do not use.
+    correctness_feedback_enabled = datastore_services.BooleanProperty(
+        default=False, indexed=True)
```
You can probably change this to default=True (just for safety).

AFZL210 replied:
Yes, this does make sense.

seanlip commented on core/templates/components/state-editor/state-editor-properties-services/state-editor.service.ts:
```diff
@@ -85,7 +85,7 @@ export class StateEditorService {
   linkedSkillId!: string;
   stateNames: string[] = [];
-  correctnessFeedbackEnabled: boolean = false;
+  correctnessFeedbackEnabled: boolean = true;
```
Why do we still need this line at all?

AFZL210 replied:
we are no longer using it. removed it.

seanlip commented on core/templates/pages/exploration-editor-page/exploration-editor-page.component.ts:
```diff
@@ -255,8 +255,7 @@ export class ExplorationEditorPageComponent implements OnInit, OnDestroy {
         explorationData.auto_tts_enabled);
-      this.explorationCorrectnessFeedbackService.init(
-        explorationData.correctness_feedback_enabled);
+      this.explorationCorrectnessFeedbackService.init(true);
```
Why do we need this line (an the entire service) at all?

AFZL210 replied:
removed all unnecessary services related to correctness feedback from the frontend, like PlayerCorrectnessFeedbackEnabledService and ExplorationCorrectnessFeedbackService, as they were only used for getting info about correctness feedback and responsible for changing it (which we don't need anymore).

seanlip commented on core/templates/pages/exploration-editor-page/preview-tab/preview-tab.component.ts:
```diff
@@ -152,8 +152,7 @@ export class PreviewTabComponent
           });
-        this.playerCorrectnessFeedbackEnabledService.init(
-          returnDict.correctness_feedback_enabled);
+        this.playerCorrectnessFeedbackEnabledService.init(true);
```
Why do we need this line (and the service in question)?

AFZL210 replied:
this was being used in exploration editor to get correctness feedback state and update it, which we dont need anymore so removed the entire service.

seanlip commented on core/templates/pages/exploration-editor-page/services/exploration-correctness-feedback.service.ts:
```diff
@@ -44,12 +44,7 @@ export class ExplorationCorrectnessFeedbackService extends
   }
 
   isEnabled(): string | boolean {
```
Why do we need this service at all?

AFZL210 replied:
Dropped the service as it was unnecessary

seanlip commented on core/templates/pages/exploration-editor-page/settings-tab/settings-tab.component.html:
```diff
@@ -269,32 +269,6 @@ <h3 tabindex="0">Advanced Features</h3>
         </div>
       </div>
       <hr class="oppia-feature-separator">
```
Do we need this hr?

AFZL210 replied:
Toggle buttons are wrapped/separated with hr tags. here also.
![image](https://github.com/oppia/oppia/assets/79896602/85565de8-9d7b-4599-a427-e4115eca48f8)

seanlip replied:
But do we need it if there's only one toggle? What does it currently look like for the part of the page that used to contain the correctness feedback setting?

AFZL210 replied:
currently it looks like this.
![image](https://github.com/oppia/oppia/assets/79896602/3d1c977e-5785-4fbc-bd35-9133a9ab2579)

after removing it looks like this.
![image](https://github.com/oppia/oppia/assets/79896602/65193773-aa18-4a47-9919-bb70f2c96ffc)

seanlip replied:
Ah ok, thanks. Either is fine, I think. Feel free to leave it as-is.

seanlip commented on core/templates/pages/exploration-player-page/services/exploration-player-state.service.ts:
```diff
@@ -160,12 +160,10 @@ export class ExplorationPlayerStateService {
       returnDict.record_playthrough_probability);
-    this.playerCorrectnessFeedbackEnabledService.init(
-      returnDict.correctness_feedback_enabled);
+    this.playerCorrectnessFeedbackEnabledService.init(true);
```
Do we still need these calls and this service?

AFZL210 replied:
Though we have removed correctness feedback from the backend completely, it makes sense to keep this in the frontend to provide more context, and it's being used in multiple places.

seanlip replied:
No, I disagree. If it's not modifiable then why do we need to have an entire service for maintaining whether we show it or not? We don't have similar services for all the other small features in the exploration player (unless they are configurable or we need to maintain state in some way).

AFZL210 replied:
understood. yes, this service is totally unnecessary now will remove it.

seanlip commented on core/templates/pages/exploration-player-page/services/exploration-player-state.service.ts:
```diff
@@ -257,8 +255,7 @@ export class ExplorationPlayerStateService {
         explorationData, null, null, null, null, [], callback);
-      this.playerCorrectnessFeedbackEnabledService.init(
-        explorationData.correctness_feedback_enabled);
+      this.playerCorrectnessFeedbackEnabledService.init(true);
```
Ditto.

AFZL210 replied:
Dropped the service as it was unnecessary

seanlip commented on core/templates/pages/exploration-player-page/services/player-correctness-feedback-enabled.service.ts:
```diff
@@ -24,7 +24,7 @@ import { downgradeInjectable } from '@angular/upgrade/static';
 })
 export class PlayerCorrectnessFeedbackEnabledService {
-  static _correctnessFeedbackIsEnabled: boolean = false;
+  static _correctnessFeedbackIsEnabled: boolean = true;
```
Why do we still need this variable (and this entire service)?

AFZL210 replied:
The service is obsolete now, so I dropped it

seanlip commented on core/templates/pages/story-editor-page/modal-templates/new-chapter-title-modal.component.ts:
```diff
@@ -168,13 +168,6 @@ export class NewChapterTitleModalComponent implements OnInit {
-      this.correctnessFeedbackDisabled = true;
-      return;
-    }
     this.correctnessFeedbackDisabled = false;
```
Why do we need this variable? (Please also check if anything is referencing it, and handle that properly.)

AFZL210 replied:
this was used for showing message when correctness feedback is disabled, which we don't need anymore.

seanlip reviewed (changes requested):
Hi @AFZL210, thanks for the PR. I think you've got most of it right, but there are still some remnants that might need to be removed.  Could you PTAL?

Note that I only commented on things that _might_ need to be removed, but if you don't think they should be removed then please respond with the counterargument (rather than removing them blindly) -- I haven't dug into the dependencies in detail yet. 

Thanks!

kevintab95 reviewed (changes requested):
Thanks @AFZL210! I have the same comments as Sean. Please assign me once you have addressed the comments.

DubeySandeep reviewed (approved):
@AFZL210 Thanks for the PR! LGTM for the 4 codeowner files. 

Also, do we need to remove [this line](https://github.com/oppia/oppia/blob/f2deedb6279dd0eed190307046eb872d23edff64/core/feconf.py#L428) from feconf.py?

AFZL210 commented:
> @AFZL210 Thanks for the PR! LGTM for the 4 codeowner files.
> 
> Also, do we need to remove [this line](https://github.com/oppia/oppia/blob/f2deedb6279dd0eed190307046eb872d23edff64/core/feconf.py#L428) from feconf.py?

I believe yes, we can also remove it as it was being used in the service and domain layers which have now been removed

lkbhitesh07 reviewed (approved):
Thanks for the changes. LGTM for the code owner files.

AFZL210 committed:
Merge branch 'oppia:develop' into fix#18481-make-correctness-feedback-enabled

chris7716 reviewed (approved):
Thanks! Approved for codeowner files.

AFZL210 committed:
removed services related to correctness feedback

AFZL210 committed:
increase test coverage

AFZL210 committed:
fix test

AFZL210 committed:
fix test

AFZL210 committed:
fix test

AFZL210 committed:
Remove the unreachable if block

AFZL210 committed:
Merge branch 'develop' into fix#18481-make-correctness-feedback-enabled

AFZL210 committed:
Merge branch 'develop' into fix#18481-make-correctness-feedback-enabled

AFZL210 committed:
fix test: add optional chaining

AFZL210 commented:
update: removed all unnecessary services and restored the cmd for correctness feedback. Will make a PR soon.

AFZL210 committed:
restore correctness_feedback_enabled cmd

AFZL210 committed:
fix linter issue

AFZL210 committed:
fix linter

AFZL210 committed:
fix test

AFZL210 committed:
fix linter

AFZL210 committed:
fix test

AFZL210 committed:
fix test

AFZL210 commented:
@seanlip  @kevintab95  PTAL. I have followed the approach used to remove the next_content_id state property and removed other unnecessary services related to correctness feedback from the frontend like PlayerCorrectnessFeedbackEnabledService and ExplorationCorrectnessFeedbackService as they were only used for getting info about correctness feedback and used for changing it (which we don't need anymore).

AFZL210 commented:
@seanlip @kevintab95 PTAL.

kevintab95 reviewed (approved):
LGTM, thanks @AFZL210!

seanlip reviewed (changes requested):
@AFZL210 You haven't replied to all my previous comments in the thread so I am not going to review this.

Please make sure that every item in the "Conversation" tab is replied to (and the changes are reflected in code) before requesting a follow-up review. Thanks.

AFZL210 commented:
@seanlip PTAL. as most of the questions were related to the services, I just replied to one comment. but now I replied to all.

seanlip commented on assets/constants.ts:
```diff
@@ -7191,11 +7191,12 @@ export default {
   "NON_METADATA_PROPERTIES": [
-    "id", "states", "next_content_id_index", "version"
+    "id", "states", "next_content_id_index", "version",
+    "correctness_feedback_enabled"
```
Why is this needed here?

AFZL210 replied:
removed it, done.

seanlip commented on core/domain/caching_services_test.py:
```diff
@@ -140,7 +140,7 @@ class CachingServicesUnitTests(test_utils.GenericTestBase):
             }
         },
-        'correctness_feedback_enabled': False,
+        'correctness_feedback_enabled': True,
```
Why is correctness_feedback_enabled needed here at all?

AFZL210 replied:
as correctness_feedbackack_enabled is present in the domain object and when creating a new exploration, it is being returned but not used by the frontend anywhere. however, it is still being used for testing and for cmd purpose, just like next_content_id_index.

seanlip replied:
@AFZL210 next_content_id_index isn't deprecated. Per @DubeySandeep, it's still in use for generating id for a content as we are not relying on content hash for generating unique id for the content. So that's why it's there.

correctness_feedback_enabled is deprecated. Please drop it from the domain object and all relevant code.

AFZL210 replied:
dropped the CMD as suggested below.

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -321,8 +324,11 @@ class ExplorationChange(change_domain.BaseChange):
+        'init_state_name', 'auto_tts_enabled',
+        'next_content_id_index', 'edits_allowed',
+        # Deprecated exploration properties.
+        DEPRECATED_EXPLORATION_PROPERTY_CORRECTNESS_FEEDBACK_ENABLED
```
Why is this here? We did not do this for other deprecated properties.

AFZL210 replied:
This is just to inform developers to avoid using this property. Other fields are marked similarly, like this:

![image](https://github.com/oppia/oppia/assets/79896602/c50cb9ae-0487-4369-a142-ccbd668b3efa)

seanlip replied:
Ah, no, you can drop this altogether. The state properties are handled differently from exp properties since they are buried in a JSON object and sometimes a migration is needed from older schema versions of the State. If an exp property is deprecated we do not need to surface it in the domain layer at all.

AFZL210 replied:
Done.

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -1401,7 +1406,7 @@ def __init__(
         self.last_updated = last_updated
         self.auto_tts_enabled = auto_tts_enabled
-        self.correctness_feedback_enabled = correctness_feedback_enabled
+        self.correctness_feedback_enabled = True
```
Why does correctness_feedback_enabled still need to be a property of the Exploration domain object? We should just drop it, right? (We didn't persist other deprecated fields at this layer.)

AFZL210 replied:
I believe other properties, like next_content_id_index still persist in the domain object. that's why I have added it.

seanlip replied:
next_content_id_index is still in use; this one is not. Please drop.

AFZL210 replied:
Done.

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -6515,7 +6516,7 @@ class ExplorationMetadataDict(TypedDict):
     param_changes: List[param_domain.ParamChangeDict]
     auto_tts_enabled: bool
-    correctness_feedback_enabled: bool
+    correctness_feedback_enabled: Optional[bool]
```
Why is this line needed at all?

AFZL210 replied:
This was just to get information about the exploration, and I believe we can remove this too.

AFZL210 replied:
Done.

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -6575,7 +6573,7 @@ def __init__(
         self.param_changes = param_changes
         self.auto_tts_enabled = auto_tts_enabled
-        self.correctness_feedback_enabled = correctness_feedback_enabled
+        self.correctness_feedback_enabled = True
```
Can we just drop the field?

AFZL210 replied:
I did drop this entirely in an old commit, but to restore CMD, I had to bring it back. We are not directly updating it. instead, it still persists in the backend only for CMD purposes and testing. It is not being used anywhere else, just like STATE_PROPERTY_WRITTEN_TRANSLATIONS_DEPRECATED and STATE_PROPERTY_NEXT_CONTENT_ID_INDEX_DEPRECATED.

seanlip replied:
Why do you need this for the CMD? Can you explain? 

It should actually be removed entirely (conceptually) so I want to understand the disconnect here.

AFZL210 replied:
I followed the approach used for next_content_id (which is not applicable for this case). here, instead of writing a separate function just for updating the correctness_feedback_enabled property in the datastore field, we instead update the domain object and save it. this is the only thing i'm confused about. can you give some hints on how should i do this without the domain object?

AFZL210 replied:
dropped the CMD as suggested below.

seanlip commented on core/domain/exp_services_test.py:
```diff
@@ -2529,7 +2499,7 @@ class ZipFileExportUnitTests(ExplorationServicesUnitTests):
 blurb: ''
 category: Algebra
-correctness_feedback_enabled: false
+correctness_feedback_enabled: true
```
The only place to change this to True is in the storage layer, since we want to be careful about how we handle existing storage models..

In all other places it should be removed entirely.

AFZL210 replied:
I do agree, but the approach followed by other fields, like next_content_id_index, is still present in the domain layer.

seanlip replied:
As mentioned above, next_content_id_index is not deprecated.

AFZL210 replied:
dropped the CMD as suggested below.

seanlip commented on core/feconf.py:
```diff
@@ -426,6 +426,7 @@ class ClassifierDict(TypedDict):
 # Whether new explorations should have correctness-feedback enabled
 # by default.
 DEFAULT_CORRECTNESS_FEEDBACK_ENABLED = True
```
Why is this still needed?

AFZL210 replied:
I did remove it after Sandeep's review, but while restoring the CMD, I noticed that the default value for next_content_id was still present and not being used. So, I brought it back.

![image](https://github.com/oppia/oppia/assets/79896602/0ad8d367-65de-4f5e-aed6-897beb428d08)

seanlip replied:
See above, next_content_id_index is not deprecated.

AFZL210 replied:
dropped the CMD as suggested below.

seanlip commented on core/templates/components/state-editor/state-responses-editor/state-responses.component.ts:
```diff
@@ -151,16 +151,11 @@ export class StateResponsesComponent implements OnInit, OnDestroy {
     return (
-      (outcome.dest === currentStateName) &&
-      outcome.labelledAsCorrect);
+      (outcome?.dest === currentStateName) &&
```
Why is the question mark needed? The outcome should be provided. If it is not then something is wrong and that should be investigated properly.

AFZL210 replied:
Sure, I will add that and write a test for it.

AFZL210 replied:
@seanlip the original condition was ```if (!outcome || !this.stateEditorService.getCorrectnessFeedbackEnabled()) { return false; }``` and in testing, this if block was hit due to correctness feedback being false.

 If we add a separate check for null or undefined outcome, then we have to add a test with null and undefined outcome, which I believe is unnecessary just for the sake of test coverage. The current ?. (optional chaining) also gives the same output if the outcome is not provided i.e return false.

seanlip replied:
Yup this seems fine -- but please add a test for null outcome if one is missing. This is important to ensure the code continues to work correctly in the future.

AFZL210 replied:
Done.

seanlip commented on core/templates/pages/exploration-editor-page/translation-tab/state-translation/state-translation.component.html:
```diff
@@ -248,7 +248,7 @@ <h3 tabindex="0">Content</h3>
                                      [showWarning]="contentIdNeedUpdates(answerGroup.outcome.feedback.contentId)"
                                      (navigateToState)="navigateToState($event)"
-                                     [correctnessFeedbackEnabled]="explorationCorrectnessFeedbackService.isEnabled()">
+                                     [correctnessFeedbackEnabled]="true">
```
Why keep this? We need to get rid of correctnessFeedbackEnabled entirely from the codebase, with very few exceptions as discussed in code review. Otherwise we're just carrying around dead code with us.

AFZL210 replied:
My bad this is totally unnecessary. I will remove it.

AFZL210 replied:
Done.

seanlip reviewed (changes requested):
@AFZL210 There are still remnants of correctness feedback in the resulting code after your PR changes.

When requesting your next review, please list the places where mention of correctness feedback remains (do a grep for "correctness" in both the frontend/backend to make sure you identify all locations). There should be very few of these. If you have questions for what to do for a particular case, please ask.

seanlip commented:
@AFZL210 Replied to each comment directly, PTAL. Thanks!

AFZL210 commented:
thanks @seanlip , I believe almost everything is done. will do changes as suggested, but the only thing i'm a little confused about is how do i bring back EditExplorationPropertyCorrectnessFeedbackEnabledCmd? as i'll drop correctness feedback from domain layer (as suggested). can you please give a high level approach to this? i did check other fields, they do use the domain layer, but they are different from this.

seanlip commented:
When retrieving commit history for an exploration, part of the data comes in the form of cmds, and after you drop the correctness_feedback_enabled property, the cmds are still there in the prod data and may arise when viewing history.

Having said that ... a closer look suggests that you might not actually need that class anyway since it is used in apply_change_list() which won't be called henceforth. So I think you are probably OK to delete it.

AFZL210 commented:
> When retrieving commit history for an exploration, part of the data comes in the form of cmds, and after you drop the correctness_feedback_enabled property, the cmds are still there in the prod data and may arise when viewing history.
> 
> Having said that ... a closer look suggests that you might not actually need that class anyway since it is used in apply_change_list() which won't be called henceforth. So I think you are probably OK to delete it.

ohk, thanks will remove it now.

AFZL210 committed:
drop EditExplorationPropertyCorrectnessFeedbackEnabledCmd

AFZL210 committed:
fix domain

AFZL210 committed:
remove extra space

AFZL210 committed:
fix editor test

AFZL210 committed:
fix: handle null and undefined outcome

AFZL210 committed:
fix state-response component

AFZL210 committed:
fix test

AFZL210 committed:
Merge branch 'develop' into fix#18481-make-correctness-feedback-enabled

AFZL210 committed:
fix exp_domain_test

AFZL210 committed:
Merge branch 'fix#18481-make-correctness-feedback-enabled' of https://github.com/AFZL210/oppia into fix#18481-make-correctness-feedback-enabled

AFZL210 committed:
fix user services test

AFZL210 committed:
update exp_domain

AFZL210 committed:
add test for null outcome

AFZL210 committed:
fix test

AFZL210 committed:
fix test

AFZL210 committed:
add undefined check

AFZL210 committed:
add ts-ignore flag

AFZL210 commented:
@seanlip PTAL

seanlip commented on core/domain/exp_services_test.py:
```diff
@@ -7078,46 +7039,6 @@ def test_update_exploration_auto_tts_enabled(self) -> None:
         self.assertEqual(exploration.title, 'new title')
         self.assertEqual(exploration.auto_tts_enabled, True)
 
-    def test_update_exploration_correctness_feedback_enabled(self) -> None:
```
@AFZL210 Could you please replace this with a similar, but differently-named, test that checks the main thing that's being tested here (i.e. that it's possible to update an older version of the exploration)? Maybe, for this, you could edit some other property instead of correctness feedback.

AFZL210 replied:
can you explain a little more on this? we already have test for other properties.

seanlip replied:
This test is not actually trying to test the property. It's trying to see if an older version of the exploration can still be updated (look at the PR which introduced it, using git blame). The author of the test just used the correctness_feedback_enabled field because it was convenient.

The request here is to keep the test but use a different field instead (and fix the name because it's misleading).

AFZL210 replied:
added test, PTAL.

seanlip commented on core/templates/components/state-editor/state-responses-editor/state-responses.component.spec.ts:
```diff
@@ -657,9 +657,21 @@ describe('State Responses Component', () => {
   });
 
+  it('should reject self-loop with null or undefined outcome', () => {
+    // This throws "TS2322". We need to suppress this error because
```
Just to check, are null and undefined valid inputs here (can they be passed in practice when the code is running in production)?

AFZL210 replied:
yes outcome can be null

screen shot without null check
![Screenshot from 2024-01-24 22-39-39](https://github.com/oppia/oppia/assets/79896602/0837f2fb-fb3a-4ebb-8e18-8ba821505907)

seanlip replied:
Here you say it can be null, but below you say outcome must be of type Outcome. Which is it?

AFZL210 replied:
eventually we get JS and outcome can be null

seanlip replied:
I think you might need to explain this more clearly. I don't understand your response.

AFZL210 replied:
I believe we are getting a null outcome because of the cloneDeep method, or am I missing something?

![image](https://github.com/oppia/oppia/assets/79896602/926c077b-76ec-45d0-851e-e134f92af92f)

AFZL210 replied:
In the ngOnInit method of the training modal component we are initializing the outcome as null, and I believe that is the reason for the null outcome
![Screenshot from 2024-01-24 23-25-34](https://github.com/oppia/oppia/assets/79896602/330874b3-1f91-4557-a057-aedd143da32f)

seanlip replied:
The training modal is not used in the exploration editor currently. I suggest digging deeper perhaps, or explaining why you think this is the cause.

Don't just poke around pieces of code randomly though. Trace backwards through the code and what calls the function in question to understand where the value originates. Feel free to start a debugging doc (see wiki) if you need help with this.

seanlip replied:
Adding lots of code with ts-ignore counts as "making things worse".

Please make the proper fix.

AFZL210 replied:
The only reason we need ts-ignore is for testing it in the undefined case. removed the null test as this will be fixed in the next PR.

seanlip commented on core/templates/components/state-editor/state-responses-editor/state-responses.component.spec.ts:
```diff
@@ -682,9 +694,7 @@ describe('State Responses Component', () => {
-    ' false if correctness feedback is not enabled', () => {
-    spyOn(stateEditorService, 'getCorrectnessFeedbackEnabled').and.returnValue(
-      false);
+    ' true if correctness feedback is enabled', () => {
```
Why are there still references to "correctnes feedback" in the test descriptions?

Please don't blindly find-and-replace -- make sure to read the tests in this file to ensure the whole suite still makes sense conceptually. If duplicate tests need to be removed, feel free to do that.

AFZL210 replied:
my bad, will remove the correctness feedback part as this should always return true because it do not depend on correctness feedback anymore.

seanlip commented on core/templates/components/state-editor/state-responses-editor/state-responses.component.ts:
```diff
@@ -151,16 +151,15 @@ export class StateResponsesComponent implements OnInit, OnDestroy {
   isSelfLoopThatIsMarkedCorrect(outcome: Outcome): boolean {
-    if (!outcome ||
-        !this.stateEditorService.getCorrectnessFeedbackEnabled()) {
+    if (outcome && typeof outcome === 'object') {
```
Why do you need the `typeof` check? It wasn't there before.

We should also be really clear in the method signature what the type of `outcome` is. It's still unclear to me whether null/undefined is a valid input or not.

AFZL210 replied:
I agree typeof is not needed as on the frontend the input is also type of Outcome and I added this as this was being followed in other functions also

seanlip replied:
typeof is stlil there, and you haven't addressed the second part of my comment.

AFZL210 replied:
Remove it, and yes undefined is a valid input because in state-responses.component, we are not initializing the value of defaultOutcome, and it is initialized in the ngOnInit function, which takes some time.

seanlip replied:
That seems problematic. Why is this function being called before ngOnInit is called? I thought ngOnInit is supposed to be the first thing called when a component is initialized.

AFZL210 replied:
Yes, ngOnInit is called after the component is initialized (the execution flow is constructor -> ngOnChanges -> ngOnInit). 

We subscribe to various events and set outcome when an event is emitted. I guess that can be a possible reason.

seanlip reviewed (changes requested):
Thanks @AFZL210. It looks pretty good! I took another pass.

AFZL210 committed:
Merge branch 'develop' into fix#18481-make-correctness-feedback-enabled

seanlip commented:
@AFZL210 Why is this PR still not yet merged? Please finish it before taking up other issues, thanks.

AFZL210 commented:
@seanlip, I was trying to understand the reason why the outcome can be undefined and null sometimes, and I think I found it.

The outcome will be null in the case of terminal interaction (EndExploration Interaction).
and outcome is undefined because in state-responses.component, we are not initializing the value of defaultOutcome, and it is initialized in the ngOnInit function, which takes some time.

undefined outcome when creating a new interaction
![image](https://github.com/oppia/oppia/assets/79896602/5758769e-0519-4c3a-9fd9-3fbc5a2988c8)


null outcome in case of EndExploration Interaction

[Screencast from 01-02-24 06:54:48 PM IST.webm](https://github.com/oppia/oppia/assets/79896602/211c7184-ed9a-4e90-8c6c-3128e91219c5)

seanlip commented:
OK, thanks, that seems plausible. Two questions:

- Can we avoid initializing the defaultOutcome in the case of an EndExploration, since we don't show it anyway? Or at least avoid calling the method in question when the default outcome is null?
- If not, does that mean that the type should be `Outcome | null` instead? (But we should try to avoid this and use the stricter typing if at all possible.)

AFZL210 commented:
I think we should let the type be Outcome | null for now, as the whole implementation is based on considering null as the outcome for EndExploration, and wherever we could have a null outcome, it is handled.

seanlip commented:
@AFZL210 I need to see a clearer analysis. Can you please explain what would need to be done exactly if we were to forbid null? Presumably we have to stop some methods from being called -- which methods are those and what calls them? Then we can decide whether it is worth doing or not.

AFZL210 commented:
1. The outcome is used in editor tab components, specifically for interactions, and for various checks like detecting loops, feedback, etc, we can create a function to check EndExploration.

something like this: 
 ``` isEndExploration(): boolean {
    if (this.stateInteractionIdService.savedMemento === 'EndExploration') {
      return true;
    } else {
      return false;
    }
  }
```
2. In [response services](https://github.com/oppia/oppia/blob/develop/core/templates/pages/exploration-editor-page/editor-tab/services/responses.service.ts#L353), only set the default outcome when the interaction type is not EndExploration.

3. When any function has the outcome as an argument, short-circuit it with the isEndExploration() function.

for example:
the outcome editor calls the isSelfLoopWithNoFeedback function, so call isEndExploration before calling it (the AND operator will not call the second function if the first one is false).

```!isEndExploration() && isSelfLoopWithNoFeedback()```

4. Remove validation for endExploration from both the [frontend](https://github.com/oppia/oppia/blob/develop/extensions/interactions/EndExploration/directives/end-exploration-validation.service.ts#L84) and [backend](https://github.com/oppia/oppia/blob/develop/core/domain/state_domain.py#L2004).

I tried implementing this and it worked, but for the end exploration, even after not setting it, I'm getting an empty object with the destination name as the state name. so need to understand why that is happening.

AFZL210 committed:
Merge branch 'develop' into fix#18481-make-correctness-feedback-enabled

AFZL210 committed:
add test

AFZL210 committed:
fix merge conflict

AFZL210 committed:
Merge branch 'develop' into fix#18481-make-correctness-feedback-enabled

AFZL210 committed:
fix conflict

seanlip commented:
I see, thanks @AFZL210 for checking. The key link is actually [this one](https://github.com/oppia/oppia/blob/develop/core/domain/state_domain.py#L2004) that you pointed to in item 4. That implies that the outcome must be nullable. So the correct approach is as follows:

- Where an outcome in general is needed for code that also handles EndExploration, its type should be null or Outcome.
- For code that should never be touched in the EndExploration case (e.g. display of outcome components in the exp editor), the type of `outcome` should be Outcome (non-nullable) if it's being passed into a function, and the execution should not reach that code in the case where the interaction is EndExploration.

Does this give you a clear path and a framework to make decisions for moving forward? Please ask if you have further questions. Do not change the validation in the backend/frontend and do not do item 2 above.

seanlip commented:
Thanks @AFZL210. I don't mind doing it in a separate PR as long as this particular PR doesn't make things "worse" (for that other PR to make better later). But if doing it separately means that this PR makes things worse, then better to sort out the issue in this PR. Does that make sense?

AFZL210 commented:
> Thanks @AFZL210. I don't mind doing it in a separate PR as long as this particular PR doesn't make things "worse" (for that other PR to make better later) -- e.g. adding a new typescript bypass or similar. But if doing it separately means that this PR makes things worse, then better to sort out the issue in this PR. Does that make sense?

Absolutely, will just do it in this PR.

AFZL210 commented:
> Thanks @AFZL210. I don't mind doing it in a separate PR as long as this particular PR doesn't make things "worse" (for that other PR to make better later) -- e.g. adding a new typescript bypass or similar. But if doing it separately means that this PR makes things worse, then better to sort out the issue in this PR. Does that make sense?

Absolutely, I think we can create a separate PR as this one doesn't really introduce anything new. 

And everything is completed in this one. You can review it, thanks!

AFZL210 committed:
Merge branch 'develop' into fix#18481-make-correctness-feedback-enabled

AFZL210 committed:
remove typeof

seanlip commented:
@AFZL210 I left a comment here: https://github.com/oppia/oppia/pull/19274#discussion_r1483243426

Also, if you are planning to punt something to a future PR, please file a bug clearly describing the issue and link to it from here, thanks.

AFZL210 committed:
Merge branch 'develop' into fix#18481-make-correctness-feedback-enabled

seanlip reviewed (approved):
OK, thanks @AFZL210 -- the new issue looks good. LGTM!

AFZL210 committed:
Merge branch 'develop' into fix#18481-make-correctness-feedback-enabled
